File: FarLib/MakeMsg.pl
#!/usr/bin/perl -w
# MakeMsg Input.msg Header.h First.lng[=cp] [Second.lng...]
use Data::Dumper;
use File::BOM qw(:all);
die "Insifficient parameters\n" if @ARGV < 3;

my $Input = shift @ARGV;
my $Header = shift @ARGV;
my @Language = map { /^(.*)=(.*)$/ ? {'file'=>$1, 'cp'=>$2} : /rus\./i ? {'file'=>$_, 'cp'=>'cp866'} : {'file'=>$_, 'cp'=>'cp850'} } @ARGV;
my @Lang = ();
my $Line;

open(HDR, "<$Header") or die "Cannot open '$Header': $!\n";
open(HDROUT, ">${Header}_out") or die "Cannot create '${Header}_out': $!\n";
while (defined($Line=<HDR>)) {
	chomp $Line;
	print HDROUT "$Line\n";
	last if $Line eq 'enum eStringTable {';
}
while (defined($Line=<HDR>)) {
	chomp $Line;
	last if $Line eq '};';
}

#open(MSG, '<:via(File::BOM)', $Input) or die "Cannot open '$Input': $!\n";
open_bom(MSG, $Input, ':encoding(cp866)');

$Line=<MSG>;
chomp $Line;
my @Fields = split("\t+", $Line);
die "Invalid first line: '$Line'\n" if shift @Fields ne ".Language";

for (my $Lng = 0; $Lng < @Fields; $Lng++) {
	my $File = $Language[$Lng]->{'file'};
	my $Enc  = '>:encoding('.$Language[$Lng]->{'cp'}.')';

	open($Lang[$Lng], $Enc, $File) or die "Cannot create '$File': $!\n";
	print {$Lang[$Lng]} ".Language=$Fields[$Lng]\n\n";
}

while (defined($Line=<MSG>)) {
#	chomp $Line;
	$Line =~ s/\s+$//;
	next if $Line =~ /^(#|\/\/|;)/;

	if ($Line =~ /^\s*$/) {
		print HDROUT "\n";
		for (my $Lng = 0; $Lng < @Lang; $Lng++) {print {$Lang[$Lng]} "\n";}
	} else {
		@Fields = split("\t+", $Line);

		print HDROUT "\t$Fields[0],\n";
		for (my $Lng = 0; $Lng < @Lang; $Lng++) {
			my $Field = $Fields[$Lng+1];
			$Field = $Fields[1] unless defined $Field;
			$Field = $Fields[0] unless defined $Field;
			print {$Lang[$Lng]} $Field, "\n";
		}
	}
}

print HDROUT "};\n";
while (defined($Line=<HDR>)) {
	print HDROUT $Line;
}
close(HDROUT);
close(HDR);
rename("${Header}_out", $Header);

close(MSG);

File: FarLib/MakeMsg.py
import sys

def split_by_tabs(line:str):
  return list([l for l in line.strip().split("\t") if len(l) > 0])

msgfile = sys.argv[1]
hdrfile = sys.argv[2]
lngfiles = sys.argv[3:]

msg = open(msgfile, encoding='utf-8-sig')
hdr = open(hdrfile, "w")
lng = list([open(f, "w", encoding='utf-8-sig') for f in lngfiles])

hdr.write("enum eStringTable {\n")

header = split_by_tabs(msg.readline())

if len(header)-1 != len(lng):
  print("Language file count does not match header")
  sys.exit(1)

for lang in range(0, len(lng)):
  lng[lang].write(header[0] + "=" + header[lang+1] + "\n\n")

for line in msg.readlines():
  values = split_by_tabs(line)
  if values:
    hdr.write("    " + values[0] + ",\n")
    for lang in range(0, len(lng)):
      item = values[lang+1] if lang+1<len(values) else values[1]
      lng[lang].write(item + "\n")
  else:
    hdr.write("\n")
    for lang in lng:
      lang.write("\n")

hdr.write("};\n")
